### ISSUE TITLE
Bounding box issue in YOLO

### ISSUE THREAD

**Issue by toobatehreem** [ultralytics/ultralytics#17949]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

The ground truth bboxes inside v8Detection loss look like this: 

Ground Truth Bounding Boxes: tensor([[[  6.6564, 109.5487,  36.5207, 139.4435],
         [132.6839,  67.0981, 162.5482,  96.9929],
         [ 21.5885, 149.0098,  51.4529, 178.9046],
         [397.2819, 109.5487, 427.1462, 139.4435],
         [523.3093,  67.0981, 553.1737,  96.9929],
         [412.2141, 149.0098, 442.0784, 178.9046],
         [  6.6564, 461.1116,  36.5207, 491.0064],
         [132.6839, 418.6610, 162.5482, 448.5558],
         [ 21.5885, 500.5728,  51.4529, 530.4677],
         [397.2819, 461.1116, 427.1462, 491.0064],
         [523.3093, 418.6610, 553.1737, 448.5558],
         [412.2141, 500.5728, 442.0784, 530.4677]]], device='cuda:0')
Ground Truth Labels: tensor([[[3.],
         [3.],
         [3.],
         [3.],
         [3.],
         [3.],
         [3.],
         [3.],
         [3.],
         [3.],
         [3.],
         [3.]]], device='cuda:0')

Whereas, the original bounding boxes for this image were: 

0 0.3577981651376147 0.5952380952380952 0.0764525993883792 0.08503401360544217
0 0.6804281345565749 0.4744897959183674 0.0764525993883792 0.08503401360544217
0 0.39602446483180426 0.7074829931972789 0.0764525993883792 0.08503401360544217

Now I have two questions:

1. Why the class label 0 for all the bounding boxes is converted to 3?
2. I am trying to compare predictions with gt bounding boxes inside v8Detection Loss function, for which I need to compare with the original boxes as I have to update the ground truth, is there a way to access the original bounding boxes (without mosaic) at the v8Detection loss function level and compare them with pred boxes?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @toobatehreem, thank you for reaching out and for your interest in Ultralytics 🚀! We appreciate you sharing your detailed question.

#### 📚 Resources
We recommend checking our [Documentation](https://docs.ultralytics.com) for insights, where you can find a range of helpful resources, including [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) and [Usage Examples](https://docs.ultralytics.com/usage/python/). These might provide context on the transformation of bounding boxes and labels during training.

If this is a 🐛 Bug Report, could you help us further by providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/)? This will allow us to better understand and investigate the behavior you're experiencing. For example, sharing a small sample of data and steps to replicate the bounding box transformations would be highly useful.

#### 💡 Suggestions
For accessing augmented vs. original bounding boxes during training/transformation, it may help to inspect the training data pipeline implementation. You can explore relevant code and logic in the repository, especially where augmentation methods (like mosaic) are applied. 

#### 🌍 Join the Community
If you're looking for real-time help or to brainstorm with others, consider joining our [Discord](https://discord.com/invite/ultralytics) 🎧. You can also participate in ongoing discussions on our [Discourse](https://community.ultralytics.com) or [Subreddit](https://reddit.com/r/Ultralytics).

#### 🔄 Upgrade Recommendation
To ensure you're running the latest features and fixes, please upgrade to the most recent version of `ultralytics`. Use the command below to update:

```bash
pip install -U ultralytics
```

We recommend verifying that your Python environment aligns with the required dependencies listed in the [requirements file](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml).

#### ✅ Status Check
You can view the status of our continuous integration tests here: <a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>. If the badge is green, all our CI tests are passing across macOS, Windows, and Ubuntu for various YOLO modes and tasks.

#### 📦 Verified Environments
To try YOLO in a verified environment with all dependencies pre-installed, you can use one of these options:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Cloud Platforms**: Google Cloud, AWS, or Docker. Refer to our [Environments Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/) for setup details.

This is an automated response to help you get started while an Ultralytics engineer reviews your issue and provides further assistance. Thank you for your patience! 😊

**Comment by Y-T-G** [comment]
What model is this?

**Comment by toobatehreem** [comment]
> What model is this?

YOLOv8

**Comment by toobatehreem** [comment]
> What model is this?

Even for images, where I have let's say 3-4 bounding boxes, the ground truth gt_bboxes variable in v8Detection loss doesn't give all the boxes. It is giving only 1 box, however, the original data had 3.

**Comment by Y-T-G** [comment]
Do you have class ID 3 in your dataset?

**Comment by toobatehreem** [comment]
> Do you have class ID 3 in your dataset?

yes 3 is there but it is a different category.

**Comment by glenn-jocher** [comment]
Under certain scenarios, labels might be remapped during data augmentation, such as mosaic or transformations applied during training. To debug this, you can try disabling mosaic augmentation (`mosaic: 0` in your training configuration) or inspect intermediate augmentation processes to trace how labels are being altered. As for comparing predictions with original bounding boxes inside the `v8Detection` loss, you would need to modify the dataloader or augmenters to retain and pass the original, unaltered bounding boxes through the pipeline.

**Comment by toobatehreem** [comment]
> Under certain scenarios, labels might be remapped during data augmentation, such as mosaic or transformations applied during training. To debug this, you can try disabling mosaic augmentation (`mosaic: 0` in your training configuration) or inspect intermediate augmentation processes to trace how labels are being altered. As for comparing predictions with original bounding boxes inside the `v8Detection` loss, you would need to modify the dataloader or augmenters to retain and pass the original, unaltered bounding boxes through the pipeline.

So, if my ground truth class 0 is mapped to class 3 during augmentations, will the class in the predictions by the model also be 3 or 0?

**Comment by toobatehreem** [comment]
> I don't it was remapped. It's just mosaic augmentation included other images in it.

But now my ground truth is showing class '3' for all the bounding boxes; however, the original image had class '0' and none of the bounding boxes has class '0' now.

**Comment by Y-T-G** [comment]
You can just plot the image and see what it looks like with cv2

**Comment by toobatehreem** [comment]
> You can just plot the image and see what it looks like with cv2

yes I plotted the 'img' which is a tensor in the batch, it is a mosaic augmented sample. Moreover, I also noticed that the classes are re-mapped differently for each batch. For example, my ground truth class id for 'person' was 0 for an image, after I checked the ground truth in the v8Detection loss, for the same image, the person id was 30 and in another image with a person, the id was 59. Does it mean that the ids are remapped differently for every batch.

**Comment by toobatehreem** [comment]
> You can just plot the image and see what it looks like with cv2

![image](https://github.com/user-attachments/assets/0e64acb5-86ca-462a-94b5-4c498cfcfd66)

for example in these two images (the samples generated by yoloworld when you start training), class id 37 is representing completely two different parts however in my original data these classes are differnt.

**Comment by Y-T-G** [comment]
Is this for YOLO-World training?

**Comment by toobatehreem** [comment]
> Is this for YOLO-World training?

yes

**Comment by toobatehreem** [comment]
> I think YOLO-World does that through RandomText augmentation to generate positive and negative texts.
> 
> https://github.com/ultralytics/ultralytics/blob/28f116682fcb69f543f6f3907c350ca9627337cf/ultralytics/data/augment.py#L2135

yes right, thank you so much!

**Comment by glenn-jocher** [comment]
You're welcome! If you have any further questions about YOLO-World or need clarification, feel free to ask. Glad to assist!
